### DISCUSSION TITLE
TextCategorizer - how to add labels into existing model?

### DISCUSSION THREAD

**Discussion by rsoika** [explosion/spaCy#6905] [Help: Coding & Implementations] [answered]
I am using a supervised training concept. In this scenario new text categories can appear also after the model was initialized.
I understand that this should now be possible with v3.0. (see: https://github.com/explosion/spaCy/issues/6748). But I did not figure out how to setup the `textcat` pipeline in the correct way. 

The following code raises an error :

    textcat = nlp.get_pipe("textcat")
    if myNewlabel not in textcat.labels:
       ner.add_label(myNewlabel)

error message:

```
 Traceback (most recent call last):
  File "/test_traindata_sport.py", line 81, in <module>
    prdnlp = datatrain.updateModel(training_data,modelfile)
  File "/git/imixs-ml/imixs-ml-spacy/imixs/core/datatrain.py", line 105, in updateModel
    textcat.add_label(_label)
  File "/.local/lib/python3.7/site-packages/spacy/pipeline/textcat.py", line 301, in add_label
    self._allow_extra_label()
  File "spacy/pipeline/trainable_pipe.pyx", line 218, in spacy.pipeline.trainable_pipe.TrainablePipe._allow_extra_label
ValueError: [E922] Component 'textcat' has been initialized with an output dimension of 3 - cannot add any more labels.
```

How can I solve this problem? Is the only solution to initialize a complete new model if a new category arises?
For the 'ner' pipeline it is not a problem if I provide new entities on an already existing model.

**Comment by svlandeg** [comment]
I'm afraid the `textcat` components is not actually resizable in v3, only the NER is. You can check this yourself by calling `pipe.is_resizable`.

**Comment by svlandeg** [comment]
This would definitely be an interesting feature to add though. I haven't looked in detail, but defining a new model and copying over the weights from the labels we had before, a bit like the NER does it, should be doable...

**Comment by rsoika** [comment]
Yes this feature would be helpful.

I wonder why all the examples about AI text classification are based on the only scenario to classify social media comments in positive/negative emotions?

Our business customers ask me if documents can be classified by departments in which they are processed?
Departments in an organisation are not fully dynamically. But changes over the time are possible. 
Following an incremental training approach I currently tell my customer that this is not possible -  at least using spaCy :-(

Or take another example from the area of social media: Lets assume you want to classify news in a social media channel by sports. You start with categories like 'Tennis', 'Football', 'Baseball'. But what if you later find a news about Chess or Motocross. You can not train this with SpaCy on an existing model.

I can't believe that I am the only one who thinks like this...? Or is there a restriction making it impossible to train a model like this? - e.g  a high number of categories?

**Reply by svlandeg** [reply]
One "hack" for dealing with a set of categories that could potentially change in the future, is to have an additional label "OTHER" in your current classifier, and then later on you could create a second classifier that deals with the "OTHER" cases and tries to label them with the new categories, but I definitely agree that this is not ideal.

On the other hand, I don't think new departments will be added on a daily basis? So perhaps when a new departement is added, you could just redefine the classifier and retrain it? If you streamline your training process, it really shouldn't be so much trouble to retrain from scratch every once in a while? (if you haven't yet - check out the new ["projects"](https://github.com/explosion/projects/tree/v3) workflow which could be well suited for that)

Just to reiterate - I do think we'll add the feature of resizing the `textcat` at one point, because it is definitely an important feature to have. PRs would be welcome too, ofcourse!

**Comment by rsoika** [comment, marked as answer]
Hi again,
I have now updated to spaCy version 3.1.1 and tried the new textcat_multilabel.

But still again I get the same error message with textcat_multilabel if I try to learn a new category:


```
  File "/home/rsoika/.local/lib/python3.7/site-packages/spacy/pipeline/textcat.py", line 299, in add_label
    self._allow_extra_label()
  File "spacy/pipeline/trainable_pipe.pyx", line 224, in spacy.pipeline.trainable_pipe.TrainablePipe._allow_extra_label
ValueError: [E922] Component 'textcat_multilabel' has been initialized with an output dimension of 2 - cannot add any more labels.
```

My code looks like this:

```
 textcat = nlp.get_pipe("textcat_multilabel")
....
 _label = 'my_new_category'
  textcat.add_label(_label)
```


And the config of my textcat_multilabel looks like this:

```
{
  "labels":[
    "invoice",
    "sports"
  ],
  "threshold":0.5
}
```

Have I overseen something?

Thanks for any help.

**Reply by adrianeboyd** [reply]
Check that your config is using `spacy.TextCatBOW.v2` / `spacy.TextCatCNN.v2` instead of `v1`? If you're using the exact same config as before, it would still be using the non-resizable `v1` versions.

You can update the architectures manually to `v2` or generate a new config with `spacy init config` with the default resizable models in v3.1.

**Reply by rsoika** [reply]
Thanks for you reply. 
Sorry for a silly question, but how can I set this version? I do the model creation programmatically with the following code:

```python
def initModel (modelPath):
    
    print("init new model ' " + modelPath + "'...")
    # Read language
    language=os.getenv('MODEL_LANGUAGE', 'en')
    # Test if the model exists
    modelExists=os.path.isdir(modelPath)
    if modelExists:
        raise Exception("model '" + modelPath  + "' already exists!")
    else:
        nlp = spacy.blank(language)  # create blank Language class
        
        if 'ner' not in nlp.pipe_names:
            print("...adding new pipe 'ner'...")
            nlp.add_pipe('ner')
            
        if 'textcat_multilabel' not in nlp.pipe_names:
            print("...adding new pipe 'textcat_multilabel'...")
            nlp.add_pipe("textcat_multilabel")

    nlp.initialize()
    print ("Model initalized!")
    nlp.to_disk(modelPath)
    return nlp
```
  

How can I set the right version?

**Reply by adrianeboyd** [reply]
Ah, I think the issue is that by default programmatically you get `spacy.TextCatEnsemble.v2`, which still isn't resizable. As an example, here's how you can specify just the `TextCatBOW` config:

```python
import spacy
from spacy.pipeline.textcat_multilabel import multi_label_bow_config
from thinc.api import Config

nlp = spacy.blank("en")
nlp.add_pipe("textcat_multilabel", config=Config().from_str(multi_label_bow_config))
assert nlp.get_pipe("textcat_multilabel").is_resizable
```

There's also a predefined `multi_label_cnn_config`. Definitely compare the performance before deciding between these options.

**Reply by rsoika** [reply]
Thanks!!! It works!
What did you mean with the performance issue? Do you have any concerns adding new labels into an existing model?
At all, I have no problems with performance in spacy - neither for training nor for analysis. 

We use the concept of continuous learning in our [imixs workflow system](https://www.imixs.org) . Collecting and processing the workflow data (usually invoices) takes much more time than giving the result to spacy for retraining the data. 

Or did you mean the performance between the `multi_label_bow_config` and the `multi_label_cnn_config`?

**Reply by adrianeboyd** [reply]
The default you had before was an ensemble with BOW + CNN, so either BOW or CNN on its own may not perform as well, and I think it's something you'd just need to test with your data. I also think BOW is a lot faster than CNN, if that's a factor for you.

**Reply by rsoika** [reply]
Hm, I tried to understand the documentation from [here](https://spacy.io/api/architectures#TextCatCNN). But the text is not very clear:

`spacy.TextCatCNN.v2` is  "...is usually less accurate than the ensemble, but runs faster." 
`spacy.TextCatBOW.v2` is "...should run much faster than the others, but may not be as accurate, especially if texts are short."

So both are faster but less accurate??? ;-) 
I would like to take the one which is more accurate because performance is not an issue for me.

**Reply by adrianeboyd** [reply]
The ensemble of BOW+CNN is slower and more accurate than either alone (no real surprise). BOW is much faster but usually less accurate than CNN. It can depend a lot on your data/task, though, so I'd recommend just trying them out.
